server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java

```diff
@@ -91,63 +118,249 @@ private void registerListeners() {
+        DiscoveryNode previousHealthNode = HealthNode.findHealthNode(event.previousState());
+        return Objects.equals(lastSeenHealthNode.get(), currentHealthNode == null ? null : currentHealthNode.getId()) == false
+            || Objects.equals(previousHealthNode, currentHealthNode) == false;
     }
 
+    @Nullable
     DiskHealthInfo getLastReportedDiskHealthInfo() {
-        return lastReportedDiskHealthInfo;
+        return lastReportedDiskHealthInfo.get();
+    }
+
+    /**
+     * This class is responsible for running the health monitoring in the configured intervals. The first execution happens
+     * upon initialization. If there is an exception, it will log it and continue to schedule the next execution unless it
+     * was a EsRejectedExecutionException which will cancel the task.
+     */
+    static class MonitoringTask implements Runnable, Scheduler.Cancellable {
+
+        private final TimeValue interval;
+        private final String executor;
+        private final Scheduler scheduler;
+        private final ClusterService clusterService;
+        private final DiskCheck diskCheck;
+        private final Client client;
+
+        private final AtomicReference<DiskHealthInfo> lastReportedDiskHealthInfo;
+        private final AtomicReference<String> lastSeenHealthNode;
+
+        private volatile boolean cancelled = false;
+        private volatile Scheduler.ScheduledCancellable scheduledTask;
+
+        MonitoringTask(
+            TimeValue interval,
+            Scheduler scheduler,
+            AtomicReference<DiskHealthInfo> lastReportedDiskHealthInfo,
+            AtomicReference<String> lastSeenHealthNode,
+            DiskCheck diskCheck,
+            ClusterService clusterService,
+            Client client
+        ) {
+            this.interval = interval;
+            this.executor = ThreadPool.Names.MANAGEMENT;
+            this.scheduler = scheduler;
+            this.lastReportedDiskHealthInfo = lastReportedDiskHealthInfo;
+            this.lastSeenHealthNode = lastSeenHealthNode;
+            this.clusterService = clusterService;
+            this.diskCheck = diskCheck;
+            this.client = client;
+            scheduledTask = scheduler.schedule(this, TimeValue.ZERO, executor);
+        }
+
+        /**
+         * Attempts to cancel execution of this task. This method has no
+         * effect if the task cancelled. If this task has not started when
+         * {@code cancel} is called, this task should never run anymore.
+         * If the task has already started, then the current execution
+         * will not be interrupted but the task will not be scheduled anymore.
+         *
+         * @return false, if the task was already cancelled; {@code true}
+         * otherwise.
+         */
+        @Override
+        public boolean cancel() {
+            final boolean alreadyCancelled = cancelled;
+            cancelled = true;
+            if (alreadyCancelled == false) {
```

Comment:
`scheduledTask.cancel()` but that refers to only one run. This cancels both the current run but also all subsequent runs, I think that's what makes it necessary. What do you think?

Reply:
the subsequent runs are cancelled by changing `cancelled` to true right? 
Which should still remain.

I was wondering if the `alreadyCancelled` dance is necessary since `scheduledTask.cancel` is idempotent. 

Could this cancel method be?
```
public boolean cancel() {
        cancelled = true;
       scheduledTask.cancel();
}
```

Reply:
This return value of this is not the same. Because `scheduledTask.cancel()` might return false, if the run went through but for us, it should return that it got cancelled. We can simplify in the following way though:
```
public boolean cancel() {
            if (cancelled) {
                return false;
            }
            scheduledTask.cancel();
            return cancel = true;
}
```
or something similar to make it a bit more clear. What do you think?

Reply:
After a short offline discussion with @andreidan we decided to keep an implementation that matches what Java does in `BaseFuture`:

> if future is NOT running return false // nothing to cancel
else cancel and return true